examples/src/main/java/org/bitcoinj/examples/ForwardingService.java

```diff
@@ -125,56 +149,41 @@ public void start() {
     /**
      * Setup the listener to forward received coins and wait
      */
-    public void waitForCoins() {
+    CompletableFuture<Transaction> waitForCoins() {
+        final CompletableFuture<Transaction> txFuture = new CompletableFuture<>();
         // We want to know when we receive money.
-        kit.wallet().addCoinsReceivedEventListener((w, tx, prevBalance, newBalance) -> {
+        final WalletCoinsReceivedEventListener listener = (w, tx, prevBalance, newBalance) -> {
             // Runs in the dedicated "user thread" (see bitcoinj docs for more info on this).
-            //
-            // The transaction "tx" can either be pending, or included into a block (we didn't see the broadcast).
-            Coin value = tx.getValueSentToMe(w);
-            System.out.println("Received tx for " + value.toFriendlyString() + ": " + tx);
-            System.out.println("Transaction will be forwarded after it confirms.");
-            // Wait until it's made it into the block chain (may run immediately if it's already there).
-            //
-            // For this dummy app of course, we could just forward the unconfirmed transaction. If it were
-            // to be double spent, no harm done. Wallet.allowSpendingUnconfirmedTransactions() would have to
-            // be called in onSetupCompleted() above. But we don't do that here to demonstrate the more common
-            // case of waiting for a block.
-            waitForConfirmation(tx);
-        });
+            txFuture.complete(tx);
+        };
+        kit.wallet().addCoinsReceivedEventListener(listener);
+        return txFuture.whenComplete((tx, err) ->
+            kit.wallet().removeCoinsReceivedEventListener(listener)
+        );
     }
 
     /**
      * Wait for confirmation on a transaction.
      * @param tx the transaction we are waiting for
      */
-    void waitForConfirmation(Transaction tx) {
-        tx.getConfidence().getDepthFuture(requiredConfirmations).whenComplete((result, t) -> {
-            if (result != null) {
-                System.out.println("Confirmation received.");
-                forwardCoins(forwardingAddress);
-            } else {
-                // This kind of future can't fail, just rethrow in case something weird happens.
-                throw new RuntimeException(t);
-            }
-        });
+    CompletableFuture<TransactionConfidence> waitForConfirmation(Transaction tx) {
```

Comment:
It would probably be better if waitForConfirmation() returned the `Transaction` since only the Tx hash is in the `TransactionConfidence` and the `Transaction` contains the `TransactionConfidence` and not the other way around.

Reply:
See PR #2547. If that is accepted/merged we can remove this method entirely.

Reply:
Work on PR #2547 will continue after this PR is merged.